### PR TITLE
Validate out of range for channelPricing prices

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13                   |
| Bug fix?        | yes                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no  |
| License         | MIT                                                          |

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation/ChannelPricing.xml [hunk 1]
```diff
@@ -27,20 +27,35 @@
                 <option name="minMessage">sylius.channel_pricing.price.min</option>
                 <option name="groups">sylius</option>
             </constraint>
+            <constraint name="LessThan">
+                <option name="value">2147483647.00</option>
```
Comment (Rafikooo) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation/ChannelPricing.xml]:
```suggestion
                <option name="value">2147483647</option>
```
Shouldn't it be an integer?

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation/ChannelPricing.xml [hunk 2]
```diff
@@ -27,20 +27,35 @@
...
+            </constraint>
         </property>
         <property name="originalPrice">
             <constraint name="Range">
                 <option name="min">0.00</option>
                 <option name="minMessage">sylius.channel_pricing.price.min</option>
                 <option name="groups">sylius</option>
             </constraint>
+            <constraint name="LessThan">
+                <option name="value">2147483647.00</option>
```
Comment (Rafikooo) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation/ChannelPricing.xml]:
```suggestion
                <option name="value">2147483647</option>
```

**Review by Rafikooo — DISMISSED** [review]

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation/ChannelPricing.xml [hunk 3]
```diff
@@ -27,20 +27,35 @@
                 <option name="minMessage">sylius.channel_pricing.price.min</option>
                 <option name="groups">sylius</option>
             </constraint>
+            <constraint name="LessThan">
```
Comment (Wojdylak) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation/ChannelPricing.xml]: The `Range` constraint has a `max` option, maybe we should use it. Or use the `Type` constraint, which as far as I know, checks the range of an integer.

File: features/product/managing_product_variants/product_variant_validation.feature [hunk 4]
```diff
@@ -27,6 +27,35 @@ Feature: Product variant validation
         Then I should be notified that price cannot be lower than 0
         And the "VODKA_WYBOROWA_PREMIUM" variant of the "Wyborowa Vodka" product should not appear in the store
 
+    @api @ui
+    Scenario: Adding a new product variant with price that is greater than max value allowed
```
Comment (NoResponseMate) [inline features/product/managing_product_variants/product_variant_validation.feature]:
```suggestion
    Scenario: Adding a new product variant with a price greater than max value allowed
```
Similar below

File: features/product/managing_product_variants/product_variant_validation.feature [hunk 5]
```diff
@@ -27,6 +27,35 @@ Feature: Product variant validation
...
+        When I want to create a new variant of this product
+        And I specify its code as "VODKA_WYBOROWA_PREMIUM"
+        And I set its price as out of range value for "United States" channel
```
Comment (NoResponseMate) [inline features/product/managing_product_variants/product_variant_validation.feature]:
```suggestion
        And I set its price to a huge number for the "United States" channel
```
Wdyt? Technically speaking everything below 0 is also out of range

**Review by Wojdylak — DISMISSED** [review]

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php [hunk 6]
```diff
@@ -84,6 +84,30 @@ public function iSetItsPriceToForChannel(int $price, ChannelInterface $channel):
         ]);
     }
 
+    /**
+     * @When I set its price as out of range value for :channel channel
+     */
+    public function iSetItsPriceAsOutOfRangeValueForChannel(ChannelInterface $channel): void
+    {
+        $this->iSetItsPriceToForChannel(2147483648, $channel);
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php]: Maybe extract the actual value to some const?

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation/ChannelPricing.xml [hunk 7]
```diff
@@ -23,24 +23,39 @@
...
+                <option name="value">2147483647</option>
+                <option name="groups">sylius</option>
+                <option name="message">sylius.channel_pricing.price.max</option>
+            </constraint>
         </property>
         <property name="originalPrice">
             <constraint name="Range">
-                <option name="min">0.00</option>
+                <option name="min">0</option>
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation/ChannelPricing.xml:36]: Judging by the changed scenario, it seems we're missing tests for validating the min of original and minimum prices.

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation/ChannelPricing.xml [hunk 8]
```diff
@@ -23,24 +23,39 @@
         </constraint>
         <property name="price">
             <constraint name="Range">
-                <option name="min">0.00</option>
+                <option name="min">0</option>
                 <option name="minMessage">sylius.channel_pricing.price.min</option>
                 <option name="groups">sylius</option>
             </constraint>
+            <constraint name="LessThan">
+                <option name="value">2147483647</option>
```
Comment (diimpp) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation/ChannelPricing.xml]: 1. Feels too much magical, I don't usually remember max int value, which is different at 32/64 bit systems.

image

and that value is actually for mysql/postgresql `int`.

2. It's a common case to change `int` to `bigint` in database to handle cryptocurrency, so this change needs to be mentioned in upgrade.
3. For people with `bigint`, overriding vendor validation requires to change validation group and re-creation of all constraints for entity, which is not DX friendly.

I would suggest to create custom **MaxIntConstraint** validator, which would be easy to override and change value from `int` to `bigint` without touching validation config.

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php [hunk 9]
```diff
@@ -33,6 +33,8 @@ final class ManagingProductVariantsContext implements Context
 {
     private const FIRST_COLLECTION_ITEM = 0;
 
+    private const HUGE_NUMBER = 2147483648;
```
Comment (diimpp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php]:
```suggestion
    private const MAX_INT = 2147483648;
    private const MAX_32_INT = 2147483648;
    private const MAX_DB_INT = 2147483648;
```

**Review by diimpp — CHANGES_REQUESTED** [review]
It's a common case to change `int` to `bigint` in database to handle cryptocurrency, so this change needs to be mentioned in upgrade.

File: src/Sylius/Bundle/CoreBundle/Resources/translations/validators.en.yml [hunk 10]
```diff
@@ -84,6 +84,7 @@ sylius:
     cart_item:
         not_available: '%itemName% does not have sufficient stock.'
         insufficient_stock: 'Insufficient stock'
+    max_integer: "Value must be less than {{ compared_value }}."
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Resources/translations/validators.en.yml]:
```suggestion
    max_integer: 'Value must be less than {{ compared_value }}.'
```

File: src/Sylius/Bundle/CoreBundle/Validator/Constraints/MaxInt.php [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+final class MaxInt extends Constraint
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Validator/Constraints/MaxInt.php]:
```suggestion
final class MaxInteger extends Constraint
```
🤔

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation/ChannelPricing.xml (hunk 8 above)
Comment (TheMilek) [reply src/Sylius/Bundle/CoreBundle/Resources/config/validation/ChannelPricing.xml]: Yeah, you are right, just added custom validator with parameter 👌🏽

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php (hunk 9 above)
Comment (TheMilek) [reply src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php]: Do we actually need that in behat?

File: src/Sylius/Bundle/CoreBundle/DependencyInjection/Configuration.php [hunk 12]
```diff
@@ -62,6 +62,7 @@ public function getConfigTreeBuilder(): TreeBuilder
                     ->setDeprecated('sylius/sylius', '1.10', 'The "%path%.%node%" parameter is deprecated and will be removed in 2.0.')
                     ->defaultFalse()
                 ->end()
+                ->integerNode('max_int_value')->defaultValue(2147483647)->end()
```
Comment (Wojdylak) [inline src/Sylius/Bundle/CoreBundle/DependencyInjection/Configuration.php:65]: This value should be declared as `const`.

File: src/Sylius/Bundle/CoreBundle/Validator/Constraints/MaxIntegerValidator.php [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+{
+    public function __construct(private int $maxInt)
+    {
+    }
+
+    public function validate(mixed $value, Constraint $constraint): void
+    {
+        Assert::isInstanceOf($constraint, MaxInteger::class);
+
+        if ($value >= $this->maxInt) {
```
Comment (Wojdylak) [inline src/Sylius/Bundle/CoreBundle/Validator/Constraints/MaxIntegerValidator.php:30]: How does it behave with string `value`?

Comment (TheMilek) [reply src/Sylius/Bundle/CoreBundle/Validator/Constraints/MaxIntegerValidator.php:30]: in that case there is type validation message at API and money validation at UI

File: src/Sylius/Bundle/CoreBundle/spec/Validator/Constraints/MaxIntegerValidatorSpec.php [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+final class MaxIntegerValidatorSpec extends ObjectBehavior
+{
+    function let(ExecutionContextInterface $context): void
+    {
+        $this->beConstructedWith(100);
+        $this->initialize($context);
+    }
+
+    function it_does_not_validate_value_less_than_max(ExecutionContextInterface $context): void
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/spec/Validator/Constraints/MaxIntegerValidatorSpec.php:30]:
```suggestion
    function it_does_nothing_if_value_is_less_than_max(ExecutionContextInterface $context): void
```

File: src/Sylius/Bundle/CoreBundle/spec/Validator/Constraints/MaxIntegerValidatorSpec.php [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        $context->buildViolation((new MaxInteger())->message)->shouldNotBeCalled();
+        $this->validate(50, new MaxInteger());
+    }
+
+    function it_validates_value_equal_to_max(
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/spec/Validator/Constraints/MaxIntegerValidatorSpec.php:36]:
```suggestion
    function it_adds_violation_if_value_is_equal_to_max(
```

File: src/Sylius/Bundle/CoreBundle/spec/Validator/Constraints/MaxIntegerValidatorSpec.php [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    ): void {
+        $constraintViolationBuilder->addViolation()->shouldBeCalled();
+        $constraintViolationBuilder->atPath(Argument::any())->willReturn($constraintViolationBuilder);
+        $constraintViolationBuilder->setParameter(Argument::cetera())->willReturn($constraintViolationBuilder);
+
+        $context->buildViolation((new MaxInteger())->message)->willReturn($constraintViolationBuilder);
+        $this->validate(100, new MaxInteger());
+    }
+
+    function it_validates_value_greater_than_max(
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/spec/Validator/Constraints/MaxIntegerValidatorSpec.php:48]:
```suggestion
    function it_adds_violation_if_value_is_greater_than_max(
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Kamil! :tada:
